[14.0][REF] account_payment_*: Use native payment object

pedrobaeza opened the pull request:
The previous approach creates manually the journal entries and does all the hard work, plus not being 100% compatible with the bank statement reconciliation widget (requiring a patch on OCB to see blue lines).

That decision made sense on the moment it was done (v9), where the native payment model (account.payment) was very limited, and wasn't able to store all the needed information for the bank transaction.
    
Now that the limitations are gone, we can get rid off this extra model, and generate instead `account.payment` records, using both the native model + methods to perform the same operations.
    
This serves also to workaround the problem found in #966.
    
All the code, views and tests of main module have been adapted to this new approach

@Tecnativa TT39832

pedrobaeza committed:
[REF+IMP] account_payment_order: Use native payments

The previous approach creates manually the journal entries and does all
the hard work, plus not being 100% compatible with the bank statement
reconciliation widget (requiring a patch on OCB to see blue lines).

That decision made sense on the moment it was done (v9), where the
native payment model (account.payment) was very limited, and wasn't able
to store all the needed information for the bank transaction.

Now that the limitations are gone, we can get rid off this extra model,
and generate instead `account.payment` records, using both the native
model + methods to perform the same operations.

This serves also to workaround the problem found in #966.

All the code, views and tests of main module have been adapted to this
new approach in this commit. Later commits will adapt the rest of the
modules of the suite, and add migration scripts to transit from the
previous approach to this new one.

TT39832

pedrobaeza committed:
[REF] account_banking_pain_base: Adapt module to native payment refactoring

pedrobaeza committed:
[REF] account_banking_sepa_credit_transfer: Adapt module to native payment refactoring

Notables changes:

- InstrId and EndToEndId tags are now filled with the journal entry ID,
  as there's no bank payment line identifier.

pedrobaeza committed:
[REF] account_banking_mandate: Adapt module to native payment refactoring

pedrobaeza committed:
[REF] account_banking_sepa_direct_debit: Adapt module to native payment refactoring

Notable changes:

- InstrId and EndToEndId tags are now filled with the journal entry ID,
  as there's no bank payment line identifier.

pedrobaeza commented:
@alexis-via @sbidoul @HaraldPanten @ValentinVinagre @carlosdauden here it's the long awaited refactoring to make use of the native `account.payment` object for handling the bank lines and the move generation. The stats are good: +205 lines added against −810 lines removed.

See individual commits in order to have better comprehension of the process.

Working on the migration scripts now.

pedrobaeza committed:
[REF] account_payment_order_vendor_email: Adapt module to native payment refactoring

carlosdauden reviewed (changes requested):
At some point we should extend the account.move method `def _get_invoice_in_payment_state(self):` to return 'in_payment' to activate the payment status "In process" at the invoices.
https://github.com/odoo/odoo/blob/e337a7b4b0da970041ea26d6fdc6e2d39eb3e5be/addons/account/models/account_move.py#L1389

pedrobaeza commented:
Indeed, but that's the next step not included in this PR.

carlosdauden reviewed (commented):
I have reviewed and tested and everything is correct, I am waiting for the migration scripts to approve

pedrobaeza committed:
[OU-ADD] account_payment_order: Migration scripts to native payment ref

This adapts the old bank.payment.line records to account.payment records
according the new approach.

HaraldPanten commented:
First tests seem to be working fine. But I'll need more time to check it properly.

Have you tested it in "semi-real" use cases?

carlosdauden reviewed (commented):
We should synchronize PR merge with changes in these modules so as not to cause malfunctions:

https://github.com/OCA/account-payment/blob/9f5e77607690da8f275ccec6270a2e8aad652145/account_payment_return_import_iso20022/models/payment_return.py#L16

https://github.com/OCA/account-reconcile/blob/2b77e6078785a08890205fa94dd7682eec102466/account_reconcile_payment_order/models/account_reconciliation_widget.py#L25-L33

pedrobaeza commented:
Migration scripts are completed and tested on a real environment with all the possible cases, as well as the features of the module. We have an slower payment order confirmation due to the full creation of the account.payment + account.move, but faster times later on "File successfully uploaded" phase.

pedrobaeza commented:
> We should synchronize PR merge with changes in these modules so as not to cause malfunctions:

Done on:

https://github.com/OCA/account-payment/pull/555
https://github.com/OCA/account-reconcile/pull/505

gdgellatly commented:
@dannyadair Can we review this one, full functional with current work.

carlosdauden commented:
Can you rebase?

pedrobaeza commented:
Rebased

pedrobaeza commented:
People, please review this for switching definitively to a more standard system and go to the next step to have in-payment states in invoices, which will be a very nice thing for our customers.

legalsylvain commented:
Thanks a lot for this refactoring pedro.

sbidoul commented:
Thanks for doing this work! I generally agree with the approach of replacing bank.payment.line with the native Odoo model.

I think there should be a major version bump on all modules.

Also it would be nice if modules that need changing following to this major breaking change to `account_payment_order` get a `setup.py` like this:

```
import setuptools

setuptools.setup(
    setup_requires=['setuptools-odoo'],
    odoo_addon={
        'depends_override': {
            'account_payment_order': 'odoo14-addon-account_payment_order>=14.0.2.0.0',
        },
    },
)
```

pedrobaeza commented:
Proceeding then with the merge with a major bump:

/ocabot merge major

HaraldPanten commented:
Thanks for this work!

LoisRForgeFlow commented:
Isn't this a bit of a risky refactor to push in a stable version?

It is affecting other extensions, like the ones at https://github.com/OCA/l10n-usa (an likely other localizations)

pedrobaeza commented:
Yes, it's not the ideal thing, but we are very late on this, and problems like #966 makes this even more needed. What are the problems? Check other adapted modules or tell me and I help you.

qgroulard commented:
Great work, thank you !

Looks like `move_option` got lost in the process.
The field is still on `account.payment.mode` but isn't used anywhere anymore.
Any plan on restoring this feature ?

pedrobaeza commented:
I'm removing that leftovers in #1015, but there can't be equivalent options, as now each `account.payment` generates one move, and you can't group several payments from different partners in one record.

hhgabelgaard commented:
Great work, thank you !
Has  this change been forward ported to 15.0?
Or is there any plans for this?
The upgrade path is kind of broken now it seems ?

pedrobaeza commented:
Yes, indeed. I want to do it for 15.0, but no time for now. Do you want to do it? I will review it. It should be pretty straight-forward.

acsonefho commented on account_banking_sepa_direct_debit/models/account_payment_order.py:
```diff
@@ -281,18 +278,19 @@ def generated2uploaded(self):
             all_mandates = abmo.browse([])
-            for bline in order.bank_line_ids:
-                if bline.mandate_id in all_mandates:
+            for payment in order.payment_ids:
```
@pedrobaeza This field `payment_ids` doesn't exist on `account.payment.order` model.
Or maybe there is a missing dependency?

pedrobaeza replied:
This PR adds such field: https://github.com/OCA/bank-payment/pull/979/files#diff-e1bae20141deca080d6db095e9874359ec1ed9d92f0b4d46aedc9f3ecea71cf0R80
